tree/treeplayer/src/TTreeProcessorMT.cxx

```diff
@@ -546,16 +576,25 @@ void TTreeProcessorMT::Process(std::function<void(TTreeReader &)> func)
       const auto clustersAndEntries = MakeClusters(treeNames, fileNames, maxTasksPerFile);
       const auto &clusters = clustersAndEntries.first[0];
       const auto &entries = clustersAndEntries.second[0];
-      auto processCluster = [&](const EntryCluster &c) {
-         auto r = fTreeView->GetTreeReader(c.start, c.end, treeNames, fileNames, fFriendInfo, fEntryList, {entries},
+      auto processCluster = [&](const EntryRange &c) {
+         auto r = fTreeView->GetTreeReader(c.first, c.second, treeNames, fileNames, fFriendInfo, fEntryList, {entries},
                                            std::vector<std::vector<Long64_t>>{});
          func(*r);
       };
       fPool.Foreach(processCluster, clusters);
    };
 
-   std::vector<std::size_t> fileIdxs(fFileNames.size());
-   std::iota(fileIdxs.begin(), fileIdxs.end(), 0u);
+   auto firstNonEmpty = 0u;       // start from the first available entry
+   if (fGlobalRange.first > 0u) { // late start makes sense only if a global start is provided
+      for (auto i = 0u; i < allClusters.size(); ++i) {
+         if (allClusters[i].empty()) {
+            firstNonEmpty = i;
+            break;
+         }
+      }
+   }
```

Comment:
this would be more readable as something like:

```cpp
const auto firstNonEmpty = std::distance(allClusters.begin(),
                                         std::find_if(allClusters.begin(), allClusters.end(), [](auto &c) { return !c.empty(); }));
```